### ISSUE TITLE
AdaScale with gradient accumulation: Sum or average of gradients?

### ISSUE THREAD

**Issue by lballes** [facebookresearch/fairscale#1062]
Hi!

I am wondering what is the intended way to implement gradient accumulation when using AdaScale. Usually, for gradient accumulation, I would use a training loop like this

    for i, (X, y) in enumerate(train_loader):
        if (i+1) % num_accum_steps == 0:
            loss = loss_fn(model(X), y) / num_accum_steps
            loss.backward()
            opt.step()
            opt.zero_grad()
        else:
            with model.no_sync():
                loss = loss_fn(model(X), y) / num_accum_steps
                loss.backward()

That is, when the optimizer step is applied, the gradient stored in `p.grad` is the **average** gradient across accumulation steps. However, this [line](https://github.com/facebookresearch/fairscale/blob/main/fairscale/optim/adascale.py#L460) in the AdaScale code makes me think that AdaScale expects `p.grad` to contain the **sum** of gradients across accumulation steps.

Could somebody clarify what's the intended way?

Thanks!

**Comment by min-xu-ai** [comment]
Sorry, I missed the question for a few days.

AdaScale treats each grad step as if it is from a separate rank. That means you can use grad-accumulation as a way to simulate multiple-gpu training with adascale.

Then, if you don't really want that, you can override the scale value with adascale's init function. I.e. if you use 8gpu and 8-step grad accumulation. Normally, adascale will think this is scale 64. But you can specify scale=8 and it should work. Let me know if it doesn't.

BTW, what are you using adascale for? (if you can share)

**Comment by min-xu-ai** [comment]
>  the gradient stored in p.grad is the average gradient across accumulation steps.

We usually use *sum" not average in p.grad (hence the name "accumulation").

**Comment by lballes** [comment]
Thanks a lot for the reply.

I don't want to use gradient accumulation to _simulate_ multi-GPU training, but in combination with it. When communication overhead becomes large, I would like to perform multiple gradient accumulation steps locally on each worker, then sync between workers and update. 

I'm confused about the sum vs average. If I understand correctly, when a model is wrapped in `DDP`, after each backward pass, the gradients in `p.grad` are **averaged** across all workers. So I would assume that the gradient accumulation should also be averaged, in order to retain the original scale of the gradient. Otherwise, if I update with a hypothetical ideal gain of `num_workers * num_accum_steps` I will overshoot since `p.grad` already contains the _sum_ of `num_accum_steps` gradients.

But that doesn't seem to square with this [line](https://github.com/facebookresearch/fairscale/blob/main/fairscale/optim/adascale.py#L460) in the code as well as your answer.

**Comment by min-xu-ai** [comment]
> when a model is wrapped in DDP, after each backward pass, the gradients in p.grad are averaged across all workers

Maybe I am mistaken, but I think DDP does a sum. Here is the DDP code:

https://github.com/pytorch/pytorch/blob/eff312f07be85508f049e5ddfe9ada5aa5df4fc4/torch/nn/parallel/distributed.py#L1103

then, in distributed, the default op is sum:

https://github.com/pytorch/pytorch/blob/c29b7865d02239d89d4407559a85a556039cb7c6/torch/distributed/distributed_c10d.py#L1318

Regardless it is DDP or grad-accumulation, it is doing a bigger batch, which to me makes sense to sum all grads from the entire batch.

**Comment by prabhuteja12** [comment]
Hi, 

Taking the code posted above 
```
for i, (X, y) in enumerate(train_loader):
    if (i+1) % num_accum_steps == 0:
        loss = loss_fn(model(X), y) / num_accum_steps    ---- averaging the loss over the #grad accum batches. 
        loss.backward()
        opt.step()
        opt.zero_grad()
    else:
        with model.no_sync():
            loss = loss_fn(model(X), y) / num_accum_steps  ---- averaging the loss over the #grad accum batches.
            loss.backward()
```


To be clear,  
- we divide the loss by the number of accumulation steps to compute the mean of the loss. (as indicated in the code above). We wouldn't do this explicitly without gradient accumulation because MSE, CE in Pytorch implement batch averaging. 
- If we are computing the mean loss, what is the [line 460](https://github.com/facebookresearch/fairscale/blob/main/fairscale/optim/adascale.py#L460) in Adascale for? 

Can you help us with specifically what we incorrectly understand here?

**Comment by min-xu-ai** [comment]
Thanks for persisting. I did some investigation from the git history. It seems that line 460 is a bug. It was originally introduced when grad accumulation was first added to adascale. I only verified it by training cifar, which as I now know is very tolerating of mistakes.

I went though the math again and I think the scaling on line 460 is incorrect and should be removed if gradients are accumulated.

In your case though, if you already have loss (hence gradient) pre-divided as your code is showing, then I think it is the following:

Say you have have 2 ranks and grad-accum = 2. That means you have g1 and g2 from rank 0 and g3 and g4 from rank 1.
Since you have scaled them, they became g1/2 and g2/2 and g3/2 and g4/2. For adascale, local_grad_sqr becomes:

(g1/2)^2 + (g2/2)^2 + (g3/2)^2 + (g4/2)^2 = (original local_grad_sqr) / 4

Then, the total_grad_sqr became:

((g1+g2+g3+g4)/2)^2 = (g1+g2+g3+g4)^2/4 = (original total_grad_sqr) / 4

So I think in that case, the line 460 scaling is still not needed.

I think with this bug, the gain is over-estimated when gradient accumulation is enabled. Is that what you are observing?

**Comment by lballes** [comment]
Thanks very much for digging into this. Highly appreciated!

I will adapt the code and run some experiments (probably towards the end of the week) and report back here.

**Comment by min-xu-ai** [comment]
👍thanks！

**Comment by min-xu-ai** [comment]
cc @mikerabbat

**Comment by min-xu-ai** [comment]
@lballes @prabhuteja12, I revisited this issue and found that we actually have unit tests that ensure grad_accum and DDP are the same. See here:

![image](https://user-images.githubusercontent.com/24926999/189494381-ad66552a-3ea6-4de6-8779-3c641b6dcc17.png)


If I remove the `total_grad_sqr = total_grad_sqr / (self._num_grads_to_accum**2)` line a bunch of unit test fails. The test data is in fact shared between ddp and grad_accum tests. Therefore, this line of normalization is in fact needed unless there is some error within the tests.

You can remove the line and see the test error yourself. The errors are:

![image](https://user-images.githubusercontent.com/24926999/189494484-454a9409-2111-4cb0-85a3-14805d43965f.png)

**Comment by min-xu-ai** [comment]
Update: thanks to @cyugao's great work, we now have a good understanding of this issue. Please see the PR. We will support both ways of doing gradient accumulation.
